### PR TITLE
Fix ellipseMode(CORNERS) and rectMode(CORNER)

### LINKED ISSUE #7289: ellipseMode(CORNERS) broken in some cases
### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [X] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [ ] Friendly errors
- [ ] Other (specify if possible)

### p5.js version

1.10.0

### Web browser and version

Safari 18.0

### Operating system

maxOS 14.7

### Steps to reproduce this

Basically `ellipseMode(CORNERS)` only seems to work, if the first point given is top left, and the coordinates are positive.

I've made the sketch below to test out different configurations of first point (P1) and second point (P2), as well as positive and negative coordinates. This is the result:

Screen Shot 2024-09-26 at 12 35 41

7 of the 16 test cases crash p5 (`IndexError`), 8 produce a wrong ellipse rendering, only 1 works as intended.

This also affects `arc()`, since it uses the same rendering code. 

Here is a PR to address this issue: #7290

### Steps:
1. Run this sketch: https://editor.p5js.org/groedl/sketches/pBpiKShRE (or see snippet below)

### Snippet:

```js
function setup() {
  createCanvas(600, 900);
  noFill();
  noLoop();
}

function ellipseCorners(x1, y1, x2, y2) {
  // Bounding box (light gray)
  rectMode(CORNERS);
  stroke(200);
  noFill();
  strokeWeight(1);
  rect(x1, y1, x2, y2);
  
  // P1, P2 dots
  stroke(150);
  strokeWeight(4);
  point(x1, y1);
  point(x2, y2);
  
  // P1, P2 labels
  fill(150);
  noStroke();
  text('P1', x1 + 5, y1);
  text('P2', x2 + 5, y2);
  
  // Ellipse
  noFill();
  stroke(0);
  strokeWeight(1);
  ellipseMode(CORNERS);
  
  try {
    ellipse(x1, y1, x2, y2);
    console.log(`Drawing (${x1}, ${y1}, ${x2}, ${y2})`);
  } catch (e) {
    console.warn(`Error drawing (${x1}, ${y1}, ${x2}, ${y2})`);
    console.error(e);
  }
}


function draw() {
  background(220);
  translate(width/2, height/2);
  
  // Coordinate System
  stroke(200);
  line(-width/2, 0, width/2, 0);
  line(0, -height/2, 0, height/2);
  stroke(0);
  
  
  // Quadrant I (Bottom Right)
  console.log("(I) BOTTOM RIGHT");
  ellipseCorners(100, 50, 200, 100);  // OK
  ellipseCorners(100, 200, 200, 150); // Error
  ellipseCorners(200, 300, 100, 250); // Error
  ellipseCorners(200, 350, 100, 400); // Error
  
  
  // Quadrant II (Bottom Left)
  console.log();
  console.log("(II) BOTTOM LEFT");
  ellipseCorners(-200, 50, -100, 100);  // Wrong ellipse
  ellipseCorners(-200, 200, -100, 150); // Error
  ellipseCorners(-100, 300, -200, 250); // Error
  ellipseCorners(-100, 350, -200, 400); // Wrong ellipse
  
  
  // Quadrant III (Top Left)
  console.log();
  console.log("(III) TOP LEFT");
  ellipseCorners(-200, -400, -100, -350); // Wrong ellipse
  ellipseCorners(-200, -250, -100, -300); // Wrong ellipse
  ellipseCorners(-100, -150, -200, -200); // Wrong ellipse
  ellipseCorners(-100, -100, -200, -50);  // Wrong ellipse
  
  
  // Quadrant IV (Top Rigth)
  console.log();
  console.log("(IV) TOP RIGHT");
  ellipseCorners(100, -400, 200, -350); // Wrong ellipse
  ellipseCorners(100, -250, 200, -300); // Wrong ellipse
  ellipseCorners(200, -150, 100, -200); // Error
  ellipseCorners(200, -100, 100, -50);  // Error
}
```

### PR DESCRIPTION
**Pull request by martinleopold**
Resolves #7289

Changes:
This mainly resolves a rendering/crashing issues with `ellipseMode(CORNERS)` affecting `ellipse()` and `arc()`; see issue above). It also fixes a related issue with using `rectMode(CORNER)` and negative widths and/or heights.

The bulk of the changes are in `modeAdjust()` which should now properly handle the modes and return a normalized bounding box. As a result a few lines of handling negative widths and heights needed to be removed from `arc()` and `_renderEllipse()`.

I've included a test sketch in test/shape-modes.html where you can cycle through all the different modes and shapes. I'm happy to update this, if somebody could direct me on how to properly test this case. I'm not sure, since it's mostly about the rendering output.

Screenshots of the change:
ellipse-fix

#### PR Checklist

- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated
- [x] [Unit tests] are included / updated

[Inline documentation]: https://github.com/processing/p5.js/blob/main/contributor_docs/inline_documentation.md
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

### CODE REVIEW

**Comment by perminder-17** [comment]
Hi, Thanks for taking up this task. I believe I have tested this and it works well. I have a concern: are you planning to write tests for it in the `test/shape-modes.html` file?

If so, I don't think that approach would work for your case. Perhaps we could write unit or visual tests instead. What's your opinion?

File: test/shape-modes.html [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            const GAP = 0.1;
+            arc(x1, y1, x2, y2, 0 + GAP/2, HALF_PI - GAP);
+            arc(x1, y1, x2, y2, HALF_PI + GAP, PI - GAP);
+            arc(x1, y1, x2, y2, PI + GAP, PI + HALF_PI - GAP);
+            arc(x1, y1, x2, y2, PI + HALF_PI + GAP, TWO_PI - GAP);
+        } else if (SHAPE === 'rect') {
+            rect(x1, y1, x2, y2);
+        }
+      } catch (e) {
+        console.warn(`Error drawing (${x1}, ${y1}, ${x2}, ${y2})`);
```
Comment (perminder-17) [inline test/shape-modes.html:118]: If you are trying to write tests for it? You could probably refer to this file, this could help you to write tests for it? 

https://github.com/processing/p5.js/blob/main/contributor_docs/unit_testing.md

Let me know if this file is not related to testing purpose.

**Comment by martinleopold** [comment]
Hi, I've added proper visual test and removed `test/shape-modes.html`. I had to update `test/node/helpers.js` (the test for `modeAdjust()` as well. `npm run test` passes all tests now. Please have a look and let me know if that works.

**Comment by martinleopold** [comment]
> 🎉 Thanks for opening this pull request! Please check out our [contributing guidelines](https://github.com/processing/p5.js/blob/main/CONTRIBUTING.md) if you haven't already. And be sure to add yourself to the [list of contributors on the readme page](https://github.com/processing/p5.js#contributors)!

@all-contributors please add @martinleopold for bug, code

File: test/unit/visual/cases/shape_modes.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+  // Modes to test (used with ellipseMode or rectMode, according to shape)
+  const MODES  = [ 'CORNERS', 'CORNER', 'CENTER', 'RADIUS' ];
+
+  for (let shape of SHAPES) {
+    visualSuite(`Shape ${shape}`, function() {
+
+      for (let mode of MODES) {
+        visualTest(`Mode ${mode}`, function(p5, screenshot) {
+          p5.createCanvas(240, 500);
```
Comment (perminder-17) [inline test/unit/visual/cases/shape_modes.js:78]: Alright, I believe creating a larger canvas might be an issue. When the tests run, a bigger canvas could make them take a bit longer due to its size. Plus, including files that are 15-20kB could make the codebase pretty heavy. Could we perhaps reduce it to 50x50? That would probably take up minimal space in bytes.

By doing this, we could include the tests one by one. For reference, you might want to check out this link: https://github.com/processing/p5.js/pull/6783/files#diff-aefa4b8e5287fbd0cdf37a6ec8d27322a4e5c893e99ec74d32053e60e2bdfd7d

Maybe by using a 50x50 size, we could adjust the parameter values of _shapeCorners accordingly?

**Review by perminder-17 — COMMENTED** [review]
Looks great so far. Just a few bits in the testing part. Thanks a lot for your work and your patience :)

File: test/unit/visual/cases/shape_modes.js (hunk 2 above)
Comment (martinleopold) [reply test/unit/visual/cases/shape_modes.js]: I think in this case using many smaller canvases instead of a few bigger ones, will actually make things slower. Because then we'll have the overhead of createCanvas() i.e. creating a rendering context much more often, and the overall codebase size will also be larger with many small PNGs because of the PNG headers.

And to thoroughly test everything, we have 4 modes, times 3 shapes, times 4 quadrants, times 4 combinations of points, equals 192 shapes to draw. I see no way around that. Bear in mind this adds 192 test for fundamental drawing operations that were previously untested.

Therefore I propose making the canvas in the tests as small as possible, while still testing all the cases. 
I have reduced the canvas size from 240x500 to 60x125 px, which reduces the total size of the generated PNGs from 223 to 12 kB.

**Review by perminder-17 — APPROVED** [review]
Looks good to me. 

Requesting @davepagurek for a round of review since I haven't used `ellipseMode()` much, so needs another round of review. Also, very thanks for your patience @martinleopold I am really sorry for the delay on reviewing this PR, also thanks for your hardwork.

File: test/unit/visual/cases/shape_modes.js (hunk 2 above)
Comment (perminder-17) [reply test/unit/visual/cases/shape_modes.js]: sounds good!

**Review by davepagurek — APPROVED** [review]
Thanks everyone!
